Add FieldExpansionIterator that replaces the ANYFIELD scanner

apmoriarty opened the pull request:
Add iterator that expands fields for literal values.

apmoriarty committed:
Add FieldExpansionIterator that replaces the ANYFIELD scanner

apmoriarty committed:
Correct bad assumption

apmoriarty committed:
formatting

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/data/parsers/ShardIndexKey.java:
should this be folded into DatawaveKey?

apmoriarty replied:
No. The datawave key general purpose and not reusable, the ShardIndex key and other key parsers like it (EventKey, FieldIndexKey, TermFrequencyKey) are lightweight and reusable.

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/lookups/FieldExpansionIndexLookup.java:
```diff
@@ -0,0 +1,168 @@
+                timedScanFuture = execService.submit(createTimedCallable(scanner));
+            } catch (Exception e) {
+                log.error("Error expanding term into discrete fields", e);
+                throw new RuntimeException(e);
```
cleanup the scanner

apmoriarty replied:
Even if an exception is thrown in the catch block, the finally statement will still execute, thus closing the scanner.

FineAndDandy replied:
If you get an exception inside the future that is true, but if you get an exception in the above method after creating the scanner, I don't see that there is a finally that will catch anything. None of the scanner methods have throws on them, but you are catching Exception, which covers a lot more than expected throwables.

apmoriarty replied:
Taking a second pass through this. The `createTimedCallable` method that handles executing the scanner will close it via `scannerFactory.close(scanner)`. I think that's appropriate because otherwise the finally block might close the scanner before the executor service has had a chance to execute the timed callable. 

I thought I had a finally block in a local commit that never made it up but that not correct. 

Just for fun I closed the scanner in a finally block for regex and bounded range and several tests blew up. I'll add comments to each lookup that scanners should never be closed here.

apmoriarty committed:
Add note about where scanners should be closed

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
close scanner in case of exception

ivakegg commented on warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java:
```diff
@@ -0,0 +1,152 @@
+    public FilterResult filter(Key k, Value v) {
+
+        // keep it simple for now
+        log.trace("tk: {}", k.toStringNoTime());
```
wrap with if (log.isTraceEnabled() as the toStringNoTime could be expensive

ivakegg commented on warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java:
```diff
@@ -0,0 +1,152 @@
+        if (date.compareTo(startDate) < 0) {
+            // advance to start date
+            log.trace("Key before start date: {} < {}", date, startDate);
+            return new FilterResult(false, AdvanceResult.NEXT);
```
is there a NEXT_CQ ?

apmoriarty replied:
Instead of NEXT_CQ maybe this should be a custom hint that triggers a seek to the start date..

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/data/parsers/ShardIndexKey.java:
```diff
@@ -0,0 +1,98 @@
+                    }
+                }
+            }
+            shard = cq.subSequence(0, cqSplit).toString();
```
is it worth setting up the datatype here as well?

apmoriarty replied:
KeyParsers are lazily parsed, omitting the datatype is by design.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/data/parsers/ShardIndexKey.java:
```diff
@@ -0,0 +1,98 @@
+                    }
+                }
+            }
+            datatype = cq.subSequence(cqSplit + 1, cq.length()).toString();
```
is it worth setting up the shard as well here?

apmoriarty replied:
KeyParsers are lazily parsed, omitting the shard is by design.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/lookups/FieldExpansionIndexLookup.java:
```diff
@@ -0,0 +1,172 @@
+    public FieldExpansionIndexLookup(ShardQueryConfiguration config, ScannerFactory scannerFactory, String term, Set<String> fields,
+                    ExecutorService execService) {
+        super(config, scannerFactory, true, execService);
+        this.term = term.toLowerCase();
```
why lower case here?  normalizers do not necessarily lower case values.

apmoriarty replied:
Right you are...I need to revisit this case. The previous implementation was doing a just-in-time expansion for normalization which implies the ExpandMultiNormalizedTerms doesn't handle unfielded terms well..

apmoriarty replied:
I went on a journey through the query planner...

ivakegg commented on warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java:
```diff
@@ -0,0 +1,152 @@
+        }
+
+        if (datatypes != null && !datatypes.contains(parser.getDatatype())) {
+            // figure out if we are before the first datatype -- seek to first datatype
```
This comment does not seem relevant here

ivakegg replied:
can we seek to the next datatype?

apmoriarty replied:
Due to the structure of the shard index keys a seek based on the datatype is never practical. I'll correct the comment.

apmoriarty replied:
Revisited this. If any datatype in the filter sorts AFTER the current datatype then we can advance via a next call. If no datatype in the filter sorts after the current datatype then no match is possible and the iterator can advance to the next field.

apmoriarty committed:
Normalization is now earlier in query planning

Unfielded terms now have every Type applied

Unfielded terms are not marked eval only when a normalization exception
is thrown

Corrected TEXT function's index query to account for differences in
normalized index values and non-normalized event values

apmoriarty committed:
QueryFieldDatatype updates are now additive to account for index expansion and composite field rewrites

apmoriarty committed:
remove leftover print statements and unused import

apmoriarty committed:
FieldExpansionIterator supports seeking to start date

apmoriarty committed:
Improve FieldExpansionIterator handling of seek/next on datatype miss

apmoriarty committed:
Must run node transform rules prior to expansion via normalizers

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java:
```diff
@@ -0,0 +1,182 @@
+
+        // if field does not match, skip to next field
+        if ((fields != null && !fields.contains(parser.getField())) || found.contains(parser.getField())) {
+            log.trace("field not in set of expansion fields, or already seen this field. advancing to next field");
```
this also needs log.isTraceEnabled()

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java:
```diff
@@ -0,0 +1,182 @@
+
+        // if field does not match, skip to next field
+        if ((fields != null && !fields.contains(parser.getField())) || found.contains(parser.getField())) {
+            log.trace("field not in set of expansion fields, or already seen this field. advancing to next field");
```
consider also adding trace enabled check here and other places in the class

apmoriarty replied:
slf4j will make sure trace is enabled before building complex objects for logging, so this method call should be safe. I will check the other instances in this class to make sure I'm using it correctly

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java:
```diff
@@ -0,0 +1,182 @@
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+        if (!range.isStartKeyInclusive()) {
+            // need to skip to next column family
+            Key skip = range.getStartKey().followingKey(PartialKey.ROW_COLFAM);
```
if the start key was null (infinite -) this will throw an NPE

apmoriarty replied:
adding some precondition checks..

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/lookups/FieldExpansionIndexLookup.java:
```diff
@@ -0,0 +1,172 @@
+
+                        // We are only returning a mapping of field name to field value, no need to
+                        // determine cardinality and such at this point.
+                        log.trace("put {}:{}", columnFamily, row);
```
should this check log level? This will be called a lot potentially?

apmoriarty replied:
can't hurt

FineAndDandy commented on warehouse/query-core/src/test/java/datawave/query/UnindexedNumericQueryTest.java:
```diff
@@ -68,7 +68,7 @@ public void testNumericTerm() throws Exception {
 
         String min = "115";
         String iowa = "'indiana'";
```
these are no longer referenced

apmoriarty committed:
updates per code review

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter

apmoriarty committed:
Merge branch 'integration' into feature/field-expansion-seeking-filter
